File: app/src/main/java/com/nityam/movsync/data/repository/RoomRepository.kt
package com.nityam.movsync.data.repository

import com.nityam.movsync.data.firebase.FirebaseSync
import com.nityam.movsync.data.model.Participant
import com.nityam.movsync.data.model.ParticipantInsert
import com.nityam.movsync.data.model.Room
import com.nityam.movsync.data.model.RoomInsert
import io.github.jan.supabase.SupabaseClient
import io.github.jan.supabase.postgrest.from

class RoomRepository(
    private val supabase: SupabaseClient,
    private val firebaseSync: FirebaseSync
) {
    suspend fun createRoom(
        hostId: String,
        displayName: String,
        fingerprint: String,
        movieName: String?,
        durationMs: Long?
    ): Room {
        repeat(10) {
            val code = RoomCodeGenerator.generate()
            if (getRoomByCode(code) == null) {
                val room = supabase.from("rooms")
                    .insert(
                        RoomInsert(
                            code = code,
                            hostId = hostId,
                            movieFingerprint = fingerprint,
                            movieName = movieName,
                            movieDurationMs = durationMs
                        )
                    ) {
                        select()
                    }
                    .decodeSingle<Room>()
                addParticipant(room.id, hostId, displayName, isHost = true, verified = true)
                firebaseSync.createRoomNode(room.code)
                firebaseSync.trackPresence(room.code, hostId, displayName, isHost = true, verified = true)
                return room
            }
        }
        error("Could not generate a unique room code")
    }

    suspend fun joinRoom(
        userId: String,
        displayName: String,
        code: String,
        fingerprint: String
    ): JoinResult {
        val room = getRoomByCode(code) ?: return JoinResult.NotFound
        if (room.movieFingerprint != fingerprint) {
            return JoinResult.FingerprintMismatch(room)
        }
        addParticipant(room.id, userId, displayName, isHost = false, verified = false)
        firebaseSync.trackPresence(room.code, userId, displayName, isHost = false, verified = false)
        return JoinResult.Joined(room)
    }

    suspend fun verifyParticipant(roomId: String, roomCode: String, userId: String) {
        val participant = supabase.from("participants")
            .select {
                filter {
                    eq("room_id", roomId)
                    eq("user_id", userId)
                }
                limit(1)
            }
            .decodeList<Participant>()
            .firstOrNull()
            ?: error("Participant not found")

        supabase.from("participants").upsert(
            ParticipantInsert(
                roomId = roomId,
                userId = userId,
                displayName = participant.displayName,
                isHost = participant.isHost,
                fingerprintVerified = true,
                fullHashVerified = participant.fullHashVerified
            )
        ) {
            onConflict = "room_id, user_id"
        }
        firebaseSync.setPresenceVerified(roomCode, userId, true)
    }

    suspend fun getRoomByCode(code: String): Room? {
        return runCatching {
            supabase.from("rooms")
                .select {
                    filter {
                        eq("code", code.uppercase())
                    }
                    limit(1)
                }
                .decodeList<Room>()
                .firstOrNull()
        }.getOrNull()
    }

    suspend fun leaveRoom(roomCode: String, roomId: String, userId: String) {
        runCatching {
            supabase.from("participants").delete {
                filter {
                    eq("room_id", roomId)
                    eq("user_id", userId)
                }
            }
        }
        firebaseSync.clearPresence(roomCode, userId)
    }

    private suspend fun addParticipant(
        roomId: String,
        userId: String,
        displayName: String,
        isHost: Boolean,
        verified: Boolean
    ) {
        supabase.from("participants").upsert(
            ParticipantInsert(
                roomId = roomId,
                userId = userId,
                displayName = displayName,
                isHost = isHost,
                fingerprintVerified = verified
            )
        ) {
            onConflict = "room_id, user_id"
        }
    }
}

sealed interface JoinResult {
    data class Joined(val room: Room) : JoinResult
    data class FingerprintMismatch(val room: Room) : JoinResult
    data object NotFound : JoinResult
}

File: web/src/room-repository.js
/**
 * RoomRepository — Port of RoomRepository.kt
 * Room CRUD operations using Supabase + Firebase.
 */

import { getSupabase } from './supabase-client.js';
import * as firebaseSync from './firebase-sync.js';
import { generateRoomCode } from './room-code-generator.js';

export async function createRoom(hostId, displayName, fingerprint, movieName, durationMs) {
  const supabase = getSupabase();

  // Try up to 10 times to generate a unique room code
  for (let i = 0; i < 10; i++) {
    const code = generateRoomCode();
    const existing = await getRoomByCode(code);
    if (!existing) {
      const { data: room, error } = await supabase
        .from('rooms')
        .insert({
          code,
          host_id: hostId,
          movie_fingerprint: fingerprint,
          movie_name: movieName || null,
          movie_duration_ms: durationMs || null,
          status: 'waiting',
        })
        .select()
        .single();

      if (error) throw error;

      // Add host as participant
      await addParticipant(room.id, hostId, displayName, true, true);

      // Create Firebase room node + track presence
      await firebaseSync.createRoomNode(room.code);
      await firebaseSync.trackPresence(room.code, hostId, displayName, true, true);

      return room;
    }
  }
  throw new Error('Could not generate a unique room code');
}

export async function joinRoom(userId, displayName, code, fingerprint) {
  const room = await getRoomByCode(code);
  if (!room) return { result: 'not_found' };

  if (room.movie_fingerprint !== fingerprint) {
    return { result: 'fingerprint_mismatch', room };
  }

  await addParticipant(room.id, userId, displayName, false, true);
  await firebaseSync.trackPresence(room.code, userId, displayName, false, true);

  return { result: 'joined', room };
}

export async function getRoomByCode(code) {
  const supabase = getSupabase();
  try {
    const { data, error } = await supabase
      .from('rooms')
      .select('*')
      .eq('code', code.toUpperCase())
      .limit(1);

    if (error) return null;
    return data && data.length > 0 ? data[0] : null;
  } catch {
    return null;
  }
}

export async function leaveRoom(roomCode, roomId, userId) {
  const supabase = getSupabase();
  try {
    await supabase
      .from('participants')
      .delete()
      .eq('room_id', roomId)
      .eq('user_id', userId);
  } catch (e) {
    // Ignore errors
  }
  await firebaseSync.clearPresence(roomCode, userId);
}

async function addParticipant(roomId, userId, displayName, isHost, verified) {
  const supabase = getSupabase();
  await supabase.from('participants').upsert({
    room_id: roomId,
    user_id: userId,
    display_name: displayName,
    is_host: isHost,
    fingerprint_verified: verified,
  });
}
